### PR TITLE
Create FailureDescribers framework, implement example for Gradle Version too low failure

### PR DESCRIPTION
**Pull request by tresat**
This is Step 2 towards addressing https://github.com/gradle/gradle/issues/24609, after https://github.com/gradle/gradle/pull/27573.

This creates a `FailureDescriber` framework, for replacing generic variant resolution failures with custom messages, and implements an example described for gradle plugin API version unavailable failures.

# Benefits

- Allows for easy registration of additional custom failure messages for specific scenarios _without any modification to `ResolutionFailureHandler` core logic. 
- Compile time guarantee of data present in the failure data types (`ResolutionFailure` and extensions) - no more `null` fields.

## Minor Cleanup
- Fixed some styling to use `style` instead of writing HTML-ish tags manually.

# Future Planned Work

Deferred to keep the size of this PR manageable - but not forgotten.

- Collapse exception hierarchy.  The hierarchy information will be available as part of the `ResolutionFailure` hierarchy, _not_ as specialized exceptions.  There is no more need for specialized exceptions.
- New exception will provide the `ResolutionFailure` as a field, so that consumers can check this data as well.
- New exception will be immutable, including potential resolutions.
- All failure handling code moved under `org.gradle.internal.component.resolution.failure` and sub-packages.

### CODE REVIEW

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/GradlePluginVariantsResolutionFailureDescriber.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public NoMatchingGraphVariantsException describeFailure(ResolutionFailure failure) {
+        Optional<String> minRequiredGradleVersion = findHigherRequiredVersionOfGradle(failure.getCandidates());
+        if (minRequiredGradleVersion.isPresent()) {
+            String message = buildPluginNeedsNewerGradleVersionFailureMsg(failure.getRequestedName(), minRequiredGradleVersion.get());
+            NoMatchingGraphVariantsException result = new NoMatchingGraphVariantsException(message);
+            suggestUpdateGradle(result, minRequiredGradleVersion.get());
+            suggestDowngradePlugin(result, failure.getRequestedName());
+            return result;
+        } else {
+            throw new IllegalStateException("Expected to find a higher required version of Gradle.");
```
Comment (big-guy) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/GradlePluginVariantsResolutionFailureDescriber.java:58]: When would we ever hit this?  I guess if none of the candidates had a GRADLE_PLUGIN_API_VERSION_ATTRIBUTE?

Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/GradlePluginVariantsResolutionFailureDescriber.java:58]: Which is why it's illegal 😉.  I need to look at cleaning this code up, this is mostly cut and paste from my first attempt doing this in the `ResolutionFailureHandler` itself.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
Superseded by #27858

**Pull request closed without merging**